[Ng Wei Jie, Jack] iP

j4ck990 opened the pull request:
# Apollo

> "If you know your tasks and Apollo, you are organised. If you only know your tasks but not Apollo, you are not organised. If you know neither your tasks nor Apollo, you are lost. " - Sun Tzu, The art of organisation. ([source](https://github.com/nus-cs2103-AY2122S2/ip/pull/48))

Apollo is your neighbourhood mama shop solution to all your problems

- Organise your tasks
- ~Complete orbital project~ _coming soon in **version 11**_
- ~To the moon~ _maybe in **version 13.2**_

For the low low price of ~$0.99~ **FREE**, you can get your own Apollo today!
All you need to do is, 

1. Download Apollo from [here](https://github.com/j4ck990/ip/releases)
2. Download your cat from [here ](https://www.google.com.sg/search?q=cat&sxsrf=APq-WBueffme3gRKGAYCSL16AVbnOsoRKw:1643972286283&source=lnms&tbm=isch&sa=X&ved=2ahUKEwiDpJOB8uX1AhWjSGwGHb3hAm8Q_AUoAXoECAEQAw&biw=1653&bih=1020&dpr=1.13)😺
3. Double click and Add your tasks
4. See your organised tasks
5. Pet the cat ( •̀ᴗ•́ )و ̑̑

## Features

- [X] Welcome
- [X] Organise tasks
- [ ] Done beautifully
- [ ] Pet the cat (ლ `Д ́ )ლ

If you have a heart, please contribute so everyone can pet their cat o(TヘTo)
```java
public class Cat {
    private static void pet() {
        // Press X to pet cat
    }
}
```

Commit:
Merge Gradle Support

j4ck990 committed:
Add greeting, implement echo and exit command

j4ck990 committed:
Change logo to Apollo

Logo still using Duke logo. Identity of bot needs to change to Apollo.

Step towards creating new bot Apollo.

j4ck990 committed:
Implement add and list commands, abstract out welcome code

Welcome code(daytime, logo) unrelated to Apollo's function.

Abstract out to make code more readable.

j4ck990 committed:
Implement mark and unmark command

j4ck990 committed:
Filename and code fix

Filename, Duke.java, still did not reflect new bot name Apollo.
Code was not very readable.

Fix towards new bot name. Improve code structure.

j4ck990 committed:
Abstract out command functionality

Command code structure in Apollo.java not flexible,
does not allow for easy future improvements.

Abstracted out into separate class for better readability and improvements.

j4ck990 committed:
Implement Todo, Deadline, Event commands

j4ck990 committed:
Implement ApolloException, add Text UI Testing and testcases (+1 squashed commits)

Squashed commits:

[45a5c1e] Add ApolloException

j4ck990 committed:
Fix Text UI Testing

EXPECTED.txt using different line endings.

Replaced line endings with same line ending found in ACUTAL.txt.

j4ck990 committed:
Implement delete command and update Text UI Testing

Text UI Testing does not check for delete command.

Add testcases for delete command.

j4ck990 committed:
Modify Commands to use enum, trim

Switch case used strings, command accepts unspecified instructions.
Leading spaces stops command from functioning.

Switch case use Command enum.
Add trim() to trim leading spaces from fullCommand.

j4ck990 committed:
Update coding convention

Switch case used indented case.
continuation indent was after punctuation.

Removed indent for case statement.
Shifted punctuation to after indent.

j4ck990 committed:
Add save command

Tasklist cannot be saved.

Save data via serialization.

j4ck990 committed:
Update gitignore

Not ignoring data folder

Added to gitignore

j4ck990 committed:
Merge branch 'master' into branch-Level-8

j4ck990 committed:
Update text-ui-test

j4ck990 committed:
Update save implementation

j4ck990 committed:
Add date time parsing

j4ck990 committed:
Merge branch 'branch-Level-7'

j4ck990 committed:
Merge branch 'branch-Level-8'

j4ck990 committed:
Add packages, reorganise code

OOP implementation lacking, packages missing.

j4ck990 committed:
Add JUnit testcases

j4ck990 committed:
Tweak UI layout

j4ck990 committed:
Add find command

j4ck990 committed:
Add JavaDocs

j4ck990 committed:
Update Coding standard

j4ck990 committed:
Merge branch 'branch-Level-9'

j4ck990 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/apollo/tasks/Task.java

j4ck990 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/apollo/Apollo.java
#	src/main/java/apollo/commands/MarkCommand.java
#	src/main/java/apollo/exceptions/ApolloIllegalArgumentException.java
#	src/main/java/apollo/tasks/Task.java
#	src/main/java/apollo/tasks/TaskList.java
#	src/main/java/apollo/tasks/Todo.java
#	src/main/java/apollo/ui/Welcome.java

j4ck990 committed:
Update merge issues

FindCommand missing javadocs.

Update javadocs.

j4ck990 committed:
Update Gradle support

lchokhoe commented on src/main/java/apollo/commands/FindCommand.java:
```diff
@@ -0,0 +1,54 @@
+     *
+     * @return Tasks containing keyword.
+     */
+    @Override
+    public String execute()  {
+        for (int i = 0; i < taskList.taskCount(); i++) {
+            if (taskList.getTaskDescription(i).toLowerCase().contains(keyword)) {
+                appendToResults(taskList.getTaskString(i));
+            }
+        }
+
+        if (index == 1) {
+            return "I could not find any related tasks. ";
+        }
+        return results.toString();
```
Great way of handling 'find' command by making full use of the StringBuilder field results 👍 
Saves the hassle of changing the results field to match "I could not find any related tasks. ".

lchokhoe commented on src/main/java/apollo/parser/Parser.java:
```diff
@@ -0,0 +1,100 @@
+        String[] args = userInput.trim().split(" ", 2);
+        String command = args[0].toLowerCase();
+
+        switch (command) {
+        case "todo":
+            String description = parseArgs(args)[0].trim();
+            return new AddCommand(description,
+                    null,
+                    Task.Type.valueOf(command.toUpperCase()));
+        case "deadline": case "event":
+            description = parseArgs(args)[0].trim();
+            LocalDateTime dateTime = parseDateTime(args);
+            return new AddCommand(description,
+                    dateTime,
+                    Task.Type.valueOf(command.toUpperCase()));
+        case "mark": case "unmark":
+            return new MarkCommand(parseIndex(args), command.equals("mark"));
+        case "delete":
+            return new DeleteCommand(parseIndex(args));
+        case "list":
+            return new ListCommand();
+        case "exit":
+            return new ExitCommand();
+        case "find":
+            description = parseArgs(args)[0].trim();
+            return new FindCommand(description);
+        default:
+            return new InvalidCommand();
+        }
```
Great use of switch-case statements to handle keywords 👍

lchokhoe commented on src/test/java/apollo/commands/InvalidCommandTest.java:
```diff
@@ -0,0 +1,15 @@
+
+public class InvalidCommandTest {
+    @Test
+    void execute() throws ApolloOutOfBoundsException {
+        Command invalid = new InvalidCommand();
+        String expected = "Apologies, I do not understand that. ";
+        assertEquals(expected, invalid.execute());
```
Would it be nice to also print a list of keywords that Apollo recognizes on top of the apology? 
_It might be helpful for new users of Apollo who would be more prone to making such input mistakes 😄_

lchokhoe commented on src/test/java/apollo/parser/ParserTest.java:
```diff
@@ -0,0 +1,122 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ParserTest {
```
Super detailed testing of the Paser class 👍

lchokhoe reviewed (commented):
Overall the code was surprisingly compact which speaks volumes of the effort you put in 👍 
I really liked to way you handled the printing of responses by Apollo as well!
The tests that you made are also really detailed which is great 😄 

The only real suggestion I have is maybe updating the responses of Apollo with considerations to the overall user experience 😅

CS2103Lerulatus commented on src/main/java/apollo/parser/Parser.java:
```diff
@@ -0,0 +1,100 @@
+        String[] args = userInput.trim().split(" ", 2);
+        String command = args[0].toLowerCase();
+
+        switch (command) {
+        case "todo":
+            String description = parseArgs(args)[0].trim();
+            return new AddCommand(description,
+                    null,
+                    Task.Type.valueOf(command.toUpperCase()));
+        case "deadline": case "event":
+            description = parseArgs(args)[0].trim();
+            LocalDateTime dateTime = parseDateTime(args);
+            return new AddCommand(description,
+                    dateTime,
+                    Task.Type.valueOf(command.toUpperCase()));
+        case "mark": case "unmark":
+            return new MarkCommand(parseIndex(args), command.equals("mark"));
+        case "delete":
+            return new DeleteCommand(parseIndex(args));
+        case "list":
+            return new ListCommand();
+        case "exit":
+            return new ExitCommand();
+        case "find":
+            description = parseArgs(args)[0].trim();
+            return new FindCommand(description);
+        default:
+            return new InvalidCommand();
```
——— For PR review ———
Might I suggest further use of OOP-styles -- bundle the case statements into those Command objects itself. An `ExitCommand` object should contain the alias `"exit"` that trigger it's execution, for example.

Most commands would benefit from having multiple aliases that trigger their execution too, such as `bye, exit, quit` all being valid and recognized commands that executes the `ExitCommand` action. Implementing these aliases (if you decide to in the future) in a switch-case would be troublesome, esp. as the project scales and expands to include future commands and features, so packing them into their respective Command object would be handy for future scaling.

CS2103Lerulatus commented on src/main/java/apollo/Storage.java:
```diff
@@ -0,0 +1,71 @@
+import apollo.exceptions.ApolloIoException;
+import apollo.tasks.TaskList;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
```
——— For PR review ———
Nice and clear use of imports

j4ck990 committed:
Add GUI prototype

j4ck990 committed:
Update GUI to FXML

j4ck990 committed:
Merge branch 'branch-Level-10'

j4ck990 committed:
Clean up unused imports

j4ck990 committed:
Added checkstyle for coding style violation detection

j4ck990 committed:
Merge branch 'branch-A-CheckStyle'

j4ck990 committed:
Refactor program messages

j4ck990 committed:
Add assertions

No assertions present in code.

Conditions of private methods are not being tested.

Adding assertions allows for testing of assumed conditions
of private methods.

Let's create assertions for relevant private methods.

j4ck990 committed:
Update code quality

Code not strictly following java code quality.

Standardise and reduce error-prone code practices.

Refactoring all subpar code quality to improve it.

Let's follow the given 'code quality' topics closely and refactor the
code. (+1 squashed commits)

Squashed commits:

[3b8fbcb] Update code quality

Code not strictly following java code quality.

Standardise and reduce error-prone code practices.

Refactoring all subpar code quality to improve it.

Let's follow the given 'code quality' topics closely and refactor the
code.

j4ck990 committed:
Add help messages for commands

No help or instructions on how to use commands.

User is not able to know what commands are available.

Add help messages and a help command.

Allows for better ease of use as the help can be called as user is
using the program.

j4ck990 committed:
Merge pull request #2 from j4ck990/branch-A-Assertions

Add assertions

No assertions present in code.

Conditions of private methods are not being tested.

Adding assertions allows for testing of assumed conditions
of private methods.

Let's create assertions for relevant private methods.

j4ck990 committed:
Merge pull request #3 from j4ck990/branch-A-CodeQuality

Update code quality

Code not strictly following java code quality.

Standardise and reduce error-prone code practices.

Refactoring all subpar code quality to improve it.

Let's follow the given 'code quality' topics closely and refactor the
code. (+1 squashed commits)

Squashed commits:

[3b8fbcb] Update code quality

j4ck990 committed:
Merge pull request #4 from j4ck990/branch-C-Help

Add help messages for commands

No help or instructions on how to use commands.

User is not able to know what commands are available.

Add help messages and a help command.

Allows for better ease of use as the help can be called as user is
using the program.

j4ck990 committed:
Merge branch 'master' of https://github.com/j4ck990/ip into branch-A-CodeQuality

j4ck990 committed:
Merge branch 'master' of https://github.com/j4ck990/ip

j4ck990 committed:
Merge branch 'branch-A-CodeQuality'

j4ck990 committed:
Bug fix apollo logo

Apollo logo not appearing on greeting.

j4ck990 committed:
Update build.gradle

j4ck990 committed:
Add Ui.png, code fixes

j4ck990 committed:
Update gui

j4ck990 committed:
Update Ui.png

j4ck990 committed:
Add user guide

j4ck990 committed:
Update user guide links

j4ck990 committed:
Update user guide links

j4ck990 committed:
Fix markdown table

j4ck990 committed:
Fix markdown table

j4ck990 committed:
Update build.grable version

j4ck990 committed:
Add release link for user guide

j4ck990 committed:
Add duke link for user guide

j4ck990 committed:
Fix naming in README.md
